internal/output/output_result.go

```diff
@@ -0,0 +1,538 @@
+	var osResults []OutputEcosystemResult
+	for ecosystem, sources := range ecosystemMap {
+		ecosystemResult := OutputEcosystemResult{
+			Name:    ecosystem,
+			Sources: sources,
+		}
+
+		if isOSEcosystem(ecosystem) {
+			ecosystemResult.IsOS = true
+			osResults = append(osResults, ecosystemResult)
+		} else {
+			ecosystemResults = append(ecosystemResults, ecosystemResult)
+		}
+	}
+
+	// Sort ecosystemResults to ensure consistent output
+	slices.SortFunc(ecosystemResults, func(a, b OutputEcosystemResult) int {
+		return cmp.Compare(a.Name, b.Name)
+	})
+
+	// Sort osResults to ensure consistent output
+	slices.SortFunc(osResults, func(a, b OutputEcosystemResult) int {
+		return cmp.Compare(a.Name, b.Name)
+	})
+
+	// Add project results before OS results
+	ecosystemResults = append(ecosystemResults, osResults...)
+
+	isContainerScanning := false
+	layers := getAllLayerInfo(ecosystemResults)
+	if len(layers) > 0 {
+		isContainerScanning = true
+	}
+	vulnTypeCount := getOutputVulnTypeCount(ecosystemResults)
+
+	return OutputResult{
+		Ecosystems:          ecosystemResults,
+		VulnTypeCount:       vulnTypeCount,
+		VulnCount:           resultCount,
+		IsContainerScanning: isContainerScanning,
+		AllLayers:           layers,
+	}
+}
+
+// processOutputSource processes a single source (lockfile or artifact) and returns an SourceResult.
+func processOutputSource(packageSource models.PackageSource) OutputSourceResult {
+	var sourceResult OutputSourceResult
+	packages := make([]OutputPackageResult, 0)
+	packageMap := make(map[string]bool)
+	ecosystemName := ""
+
+	for _, vulnPkg := range packageSource.Packages {
+		if ecosystemName == "" {
+			ecosystemName = vulnPkg.Package.Ecosystem
+		}
+		key := vulnPkg.Package.Name + ":" + vulnPkg.Package.Version
+		if _, exist := packageMap[key]; exist {
+			// In container scanning, the same package (same name and version) might be found multiple times
+			// within a single source. This happens because we use the upstream source name instead of
+			// the Linux distribution package name.
+			continue
+		}
+		packageResult := processOutputPackage(vulnPkg)
+		packages = append(packages, packageResult)
+		packageMap[key] = true
+
+		sourceResult.VulnCount = updateVulnCount(sourceResult.VulnCount, packageResult.VulnCount)
+		if len(packageResult.CalledVulns) != 0 {
+			sourceResult.PackageTypeCount.Called += 1
+		}
+		// A package can be counted as both called and uncalled if it has both called and uncalled vulnerabilities.
+		if len(packageResult.UncalledVulns) != 0 {
+			sourceResult.PackageTypeCount.Uncalled += 1
+		}
+	}
+	// Sort packageResults to ensure consistent output
+	slices.SortFunc(packages, func(a, b OutputPackageResult) int {
+		return cmp.Or(
+			cmp.Compare(a.Name, b.Name),
+			cmp.Compare(a.InstalledVersion, b.InstalledVersion),
+		)
+	})
+	sourceResult.Name = packageSource.Source.String()
+	sourceResult.Packages = packages
+	sourceResult.Ecosystem = ecosystemName
```

Comment:
I'm wondering if we even need to copy Ecosystem here, can we just get the ecosystem from the first package entry to be explicit about where this ecosystem is coming from.